File: IntroToF#/src/CSharp/Immutability.cs
namespace CSharp
{
    public class Immutability
    {
        public void Sample()
        {
            Credentials credentials = new Credentials("someuser", "$3cr3t");

            // Cant do it, this class is immutable!
            credentials.Username = "anotheruser";

            // But we can create a new one
            credentials = credentials.ChangeUsername("anotheruser");
        }

        // This type is immutable
        public class Credentials
        {
            public readonly string Username;
            public readonly string Password;

            public Credentials(string username, string password)
            {
                Username = username;
                Password = password;
            }

            public Credentials ChangeUsername(string username)
            {
                return new Credentials(username, Password);
            }
        }
    }
}

File: IntroToF#/src/FSharp/Immutability.fs
module Immutability

    // Type definition
    type Credentials = { Username : string; Password : string }

    // Value binding
    let someUser = { Username = "someuser"; Password = "$3cr3t" }

    // Sad trombone! 

    someUser.Username <- "soemthingelse"

    someUser <- { Username = "anotheruser"; Password = "$3cr3t" }
    
    let someUser = { Username = "anotheruser"; Password = "$3cr3t" }

    // "But I want to change it, waaaaah!" No problem, create a new one.

    let newUser = { someUser with Username = "anotheruser" }